src/metric/rank_metric.cc

```diff
@@ -139,72 +137,77 @@ struct EvalAMS : public Metric {
   float ratio_;
 };
 
-/*! \brief Area Under Curve, for both classification and rank */
+/*! \brief Area Under Curve, for both classification and rank computed on CPU */
 struct EvalAuc : public Metric {
  private:
   template <typename WeightPolicy>
   bst_float Eval(const HostDeviceVector<bst_float> &preds,
                  const MetaInfo &info,
-                 bool distributed) {
-    CHECK_NE(info.labels_.Size(), 0U) << "label set cannot be empty";
-    CHECK_EQ(preds.Size(), info.labels_.Size())
-        << "label size predict size not match";
-    std::vector<unsigned> tgptr(2, 0);
-    tgptr[1] = static_cast<unsigned>(info.labels_.Size());
-
-    const std::vector<unsigned> &gptr = info.group_ptr_.empty() ? tgptr : info.group_ptr_;
-    CHECK_EQ(gptr.back(), info.labels_.Size())
-        << "EvalAuc: group structure must match number of prediction";
-    const auto ngroup = static_cast<bst_omp_uint>(gptr.size() - 1);
+                 bool distributed,
+                 const std::vector<unsigned> &gptr) {
+    const auto ngroups = static_cast<bst_omp_uint>(gptr.size() - 1);
     // sum of all AUC's across all query groups
     double sum_auc = 0.0;
     int auc_error = 0;
-    // each thread takes a local rec
-    std::vector<std::pair<bst_float, unsigned>> rec;
-    const auto& labels = info.labels_.HostVector();
-    const std::vector<bst_float>& h_preds = preds.HostVector();
-    for (bst_omp_uint group_id = 0; group_id < ngroup; ++group_id) {
-      rec.clear();
-      for (unsigned j = gptr[group_id]; j < gptr[group_id + 1]; ++j) {
-        rec.emplace_back(h_preds[j], j);
-      }
-      XGBOOST_PARALLEL_STABLE_SORT(rec.begin(), rec.end(), common::CmpFirst);
-      // calculate AUC
-      double sum_pospair = 0.0;
-      double sum_npos = 0.0, sum_nneg = 0.0, buf_pos = 0.0, buf_neg = 0.0;
-      for (size_t j = 0; j < rec.size(); ++j) {
-        const bst_float wt
-          = WeightPolicy::GetWeightOfSortedRecord(info, rec, j, group_id);
-        const bst_float ctr = labels[rec[j].second];
-        // keep bucketing predictions in same bucket
-        if (j != 0 && rec[j].first != rec[j - 1].first) {
-          sum_pospair += buf_neg * (sum_npos + buf_pos *0.5);
-          sum_npos += buf_pos;
-          sum_nneg += buf_neg;
-          buf_neg = buf_pos = 0.0f;
+    const auto& labels = info.labels_.ConstHostVector();
+    const auto &h_preds = preds.ConstHostVector();
+
+    #pragma omp parallel reduction(+:sum_auc, auc_error) if (ngroups > 1)
+    {
+      // Each thread works on a distinct group and sorts the predictions in that group
+      PredIndPairContainer rec;
+      #pragma omp for schedule(static)
+      for (bst_omp_uint group_id = 0; group_id < ngroups; ++group_id) {
+        // Same thread can work on multiple groups one after another; hence, resize
+        // the predictions array based on the current group
+        rec.resize(gptr[group_id + 1] - gptr[group_id]);
```

Comment:
Is `rec` being shared by threads or do they each have a copy?

Reply:
they each have a copy. it is inside the parallel block.

Reply:
Can we put it into the thread block instead?

Reply:
it is already inside the thread block. do you mean inside the for loop? it was done this way to prevent repeated (de)allocations. the resizes for the most of the groups (when threads are reused) shall be no-op (as it would *mostly* be sized appropriately)